aten/src/ATen/native/Distributions.cpp

```diff
@@ -628,12 +628,19 @@ Tensor& multinomial_out(const Tensor& self,
     // s = argmax( p / (-log(eps)) ) where eps ~ U(0, 1).
     // We can also simplify the formula above by
     // s = argmax( p / q ) where q ~ Exp(1)
-    Tensor q = at::empty_like(self).exponential_(1, std::move(gen));
+
     // In theory the probability to generate 0 from exponential distribution is
     // 0. However, on CUDA side there is a protection to avoid 0s, but on CPU
     // side, there is a very low probability to generate 0 from
-    // exponential<double>. The probability is about 2^(-DBL_MANT_DIG). We just
-    // ignore it here, but there may be some risk to get invalid output on CPU.
+    // exponential<double>. The probability is about 2^(-DBL_MANT_DIG).
+    // On CPU, MKL exponential distribution kernel is defined for q >= 0.
+    // (i.e., CPU MKL exponential variate, q, does not exclude 0)
+    // To ensure that the denominator should not be zero when q = 0.0,
+    // we add a small constant, c, to the denominator.
+    // s =  argmax( p / (q+c) ) where c > 0.
+    // Here we use c=1.0 for conveninence.
+
```

Comment:
@min-jean-cho this is a very clever move to solve this issue :)

A few suggestions:
* since CUDA has done protection, adding 1.0 won't be needed for CUDA device. Otherwise it is an additional overhead.
* we may have other issues related to this one, for example, the one from https://github.com/pytorch/pytorch/issues/101620. The question is that can we apply the same trick it ?

Also can we come up with an solution that fix `exponential_`, otherwise i am afraid there might be other issues popping out in the future with operators which rely on `exponential_`.

Reply:
echo @mingfeima that we'd better explore fixes from the `exponential_` side instead of working around the caller side. Is `torch.exponential_` supposed to exclude zeros?

Reply:
Yes, it's supposed to exclude zeros

Reply:
I think we have two options:

1. As @mingfeima suggested in offline discussion, re-sample in the unlikely event exactly zero is sampled. 

- I'm not sure if this will negatively affect performance in the unlikely event exactly zero is sampled. The probability that the sample includes exactly zero increases as lambda (exponential parameter, lambda) increases. 

2. Shifting the MKL exponential distribution location by just adding a very small constant. 

- I believe this is OK to do. If `X ~ Exp(lambda)`, then `E(X) = 1/lambda` and `V(X) = 1/lambda**2`. If you linearly transform like `Y = X + c`, where `c ~= 0`, then the distribution of `Y` is very similar to `X`. _But_ the expected value slightly changes, `E(Y) = E(X + c) = E(X) + c = (1/lambda) + c`. Variance remains the same, `V(Y) = 1/lambda**2`. If `c` is very small (e.g., c = 10**(-6), etc), the two distributions are indistinguishable. That means they are _almost_ identical. 
- I think the con of this is, it won't affect performance.

What do you think?

Reply:
IMO, as I said in https://github.com/pytorch/pytorch/issues/48841#issuecomment-1551536179, we should go with the correct one. The perf hit is minimal and since it's amortised it's basically free. You can even surround it with `C10_UNLIKELY` so that the perf hit in the common case is literally zero.

Reply:
I'm concerned about the performance impact and variance of the resampling approach. I guess we can have a benchmark to evaluate the impact.

Reply:
Preliminary performance comparison of [method 1: iterative resample when sampled points include 0.0](https://github.com/pytorch/pytorch/pull/101720/commits/7615b4d6afa48bbc8ee6048c922f795539e9d86e) vs. [method 2: shift exponential D location](https://github.com/pytorch/pytorch/pull/101720/commits/a700ccc5a37ab619ab29f5315f7f3f0f314ea4db):
Tested on 28 physical cores/socket, 1 socket on Skylake. 
|              	|                         avg time (ms)                        	|                                        	|
|--------------	|:------------------------------------------------------------:	|:--------------------------------------:	|
| **lambda**       	| **method 1: iterative resample when sampled points include 0.0** 	| **method 2: shift exponential D location** 	|
| 10 ** 0      	| 318.66                                                       	| 238.5                                  	|
| 10 ** 5      	| 318.13                                                       	| 249.03                                 	|
| 10 ** 10     	| 318.42                                                       	| 248.83                                 	|
| 10 ** 15     	| 318.35                                                       	| 238.46                                 	|
| 10 ** 20     	| 318.34                                                       	| 249.17                                 	|
| 10 ** 25     	| 318.47                                                       	| 248.43                                 	|
| 10 ** 30     	| 318.32                                                       	| 238.45                                 	|
| 10 ** 35     	| 318.46                                                       	| 248.63                                 	|
| **10 ** 40** 	| **427.02**                                                   	| **248.67**                             	|
| **10 ** 45** 	| **4579.04**                                                  	| **238.27**                             	|

avg time(ms): avg time taken for `_exponential(lambda, n=1000000000)`

Takeaways:

- [method 1](https://github.com/pytorch/pytorch/pull/101720/commits/7615b4d6afa48bbc8ee6048c922f795539e9d86e) generally always performs worse than [method 2](https://github.com/pytorch/pytorch/pull/101720/commits/a700ccc5a37ab619ab29f5315f7f3f0f314ea4db), even for lambda = 1
- for reasonable λ (e.g., < 100), the probability of observing zero will be very low and far less than 0.01% of your sample will contain zeros
- number of iterations required to remove zeros when using [method 1](https://github.com/pytorch/pytorch/pull/101720/commits/7615b4d6afa48bbc8ee6048c922f795539e9d86e) iterative resampling approach
    - in order to remove all zeros `i ≤ ⌈e**(λc) ⋅ log(n)⌉` iteration is required
    `i`: number of iterations to remove zeros
    `λ`: exponential distribution parameter lambda
    `c`: x=x if x >c, x=0* if x <= c 
    `n`: number of samples
    - at most one iteration is enough to remove all zeros for reasonable values of λ
    - for a very large λ (e.g., between 10 ** 40 and 10 ** 45, which is unreasonably large), you need more iterations to resample, and hence the worser performance (e.g., assuming c=10 ** -45, when λ=10 ** 45, `i ≤ ⌈e**(10 ** 45 ⋅ 10 ** -45) ⋅ log(n)⌉`  => `i <= ⌈e ⋅ log(n)⌉`
        -  but note that such large λ is not used for exponential distribution. If λ is huge, the distribution would look like a peak rather than exponentially decreasing (e.g., If λ = 10 ** 45, `E(X) = 1/λ = 10 ** -45` and `V(X) = 1/λ ** 2 = 10 ** -90 ~= 0`). V(X) = 0 means X is not a random variable but a constant

Reply:
Thanks @min-jean-cho for the evaluation. So, let's proceed with method 2?

Reply:
Thanks @jgong5 , yes proceed with method 2.